ts/nni_manager/core/nniExperimentsManager.ts

```diff
@@ -0,0 +1,160 @@
+        }))).filter(crashedInfo => crashedInfo.isCrashed);
+        if (updateList.length > 0){
+            const result = await this.withLock(this.updateAllStatus, updateList.map(crashedInfo => crashedInfo.experimentId), fileInfo.mtime);
+            if (result !== undefined) {
+                return result;
+            } else {
+                return this.getExperimentsInfo();
+            }
+        } else {
+            return JSON.parse(JSON.stringify(Object.keys(experimentsInformation).map(key=>experimentsInformation[key])));
+        }
+    }
+
+    public setExperimentPath(newPath: string): void {
+        if (newPath[0] === '~') {
+            newPath = path.join(os.homedir(), newPath.slice(1));
+        }
+        if (!path.isAbsolute(newPath)) {
+            newPath = path.resolve(newPath);
+        }
+        this.log.info(`Set new experiment information path: ${newPath}`);
+        this.experimentsPath = newPath;
+    }
+
+    public setExperimentInfo(experimentId: string, key: string, value: any): void {
+        try {
+            if (this.profileUpdateTimer[key] !== undefined) {
+                clearTimeout(this.profileUpdateTimer[key]);
+                this.profileUpdateTimer[key] = undefined;
+            }
+            this.withLockSync(() => {
+                const experimentsInformation = JSON.parse(fs.readFileSync(this.experimentsPath).toString());
+                if (experimentsInformation[experimentId]) {
+                    experimentsInformation[experimentId][key] = value;
+                    fs.writeFileSync(this.experimentsPath, JSON.stringify(experimentsInformation));
+                } else {
+                    this.log.error(`Experiment Manager: Experiment Id ${experimentId} not found, this should not happen`);
+                }
+            });
+        } catch (err) {
+            this.log.error(err);
+            this.log.debug(`Experiment Manager: Retry set key value: ${experimentId} {${key}: ${value}}`);
+            this.profileUpdateTimer[key] = setTimeout(this.setExperimentInfo.bind(this), 100, experimentId, key, value);
+        }
+    }
+
+    private async withLock (func: Function, ...args: any): Promise<any> {
+        return withLock(func.bind(this), this.experimentsPath, {stale: 2 * 1000, retries: 100, retryWait: 100}, ...args);
+    }
+
+    private withLockSync (func: Function, ...args: any): any {
+        return withLockSync(func.bind(this), this.experimentsPath, {stale: 2 * 1000}, ...args);
+    }
+
+    private async readExperimentsInfo(): Promise<FileInfo> {
+        return this.withLock((path: string) => {
+                const buffer: Buffer = fs.readFileSync(path);
+                const mtime: number = fs.statSync(path).mtimeMs;
+                return {buffer: buffer, mtime: mtime};
+            },
+            this.experimentsPath);
+    }
+
+    private async checkCrashed(expId: string, pid: number): Promise<CrashedInfo> {
+        const alive: boolean = await isAlive(pid);
```

Comment:
you said different experiments may use the same pid, is there any plan to fix this issue?

Reply:
not different experiments may use the same pid. Some other process may use the crashed experiment pid recorded in `.experiment`, so when we check if pid is alive to check if experiment is alive, it may make some mistake. An easy way to fix this issue is to check if the `experiment id` running on the port matches the `experiment id` recorded in `.experiment`.